# Grouping_Aggregating_Data

### This script contains the following points:

#### 01. Import Libraries and Files

#### 02. Examining the df_ords_prods_merge *(merged orders_products_prior.csv (original) and orders_checked.pkl (cleaned) and products_checked.pkl (cleaned) files)* dataframe

#### 03. Grouping and Aggregating Data

##### Calculate the mean of the order_number column grouped by department_id

##### Create a customer loyalty flag based on the mean order number

##### Calculate the statistics for the prices column grouped by the loyalty column

##### Create a spending flag for the two different customer_average_prices groups using the iloc function

##### Create a frequency flag for the three different customer_median_prior_order_days groups using the iloc function

* Exploring the NaN values and creating a frequency flag for them

#### 04. Exporting dataframe as pickle file 'orders_products_merged_grouped.pkl'

## 01. Import Libraries and Files

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Assign the main project folder path to the variable path and view
path = r'/Users/elsaekevall/Jupyter_Notebook/Career_Foundry/06_2022_Instacart_Basket_Analysis/'
path

'/Users/elsaekevall/Jupyter_Notebook/Career_Foundry/06_2022_Instacart_Basket_Analysis/'

In [3]:
# Use the os.path.join() function to import the orders_products_merged_derived.pkl file as a pandas dataframe
df_ords_prods_merge_grp = pd.read_pickle(os.path.join(path, '02_Data', '02_2_Prepared_Data', 'orders_products_merged_derived.pkl'))

## 02. Examining the df_ords_prods_merge_der *(orders_products_merged_derived.pkl file)* dataframe

In [4]:
# View the first five rows of the dataframe df_ords_prods_merge_grp
df_ords_prods_merge_grp.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc     busiest_day     Busiest_days  \
0  Mid-range product  Regularly busy     Regular days   
1  Mid-range product  Regularly busy  Least busy days   
2  Mid-range product  Regularly busy  Least busy days   
3  Mid-range product      Least busy  Least busy days   
4  Mid-range product      Least busy  Least busy days   

  busiest_period_of_the_day  
0            Average orders  
1            Average orders  
2               Most orders  
3            Average orders  
4               Most orders

In [5]:
# View the shape of the dataframe df_ords_prods_merge_grp
df_ords_prods_merge_grp.shape

(32404859, 19)

## 03. Grouping and Aggregating Data

1. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.**See steps 01. and 02. above**<br>
2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.
3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

###  Calculate the mean of the order_number column grouped by department_id

In [6]:
# Calculate the mean of the “order_number” column grouped by the “department_id” column
df_ords_prods_merge_grp.groupby('department_id').agg({'order_number': ['mean']}).round(2)

order_number
                      mean
department_id             
1                    15.46
2                    17.28
3                    17.17
4                    17.81
5                    15.22
6                    16.44
7                    17.23
8                    15.34
9                    15.90
10                   20.20
11                   16.17
12                   15.89
13                   16.58
14                   16.77
15                   16.17
16                   17.67
17                   15.69
18                   19.31
19                   17.18
20                   16.47
21                   22.90

#### With all the departments in the  dataframe listed there are more (21 compared to 8 in the subset) and the means range from 15 to 23 (11 to 19 in the subset). As expected the means for the departments in the subset which only looks at the first 1,000,000 rows are not the same as the whole dataframe.

4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

###  Create a customer loyalty flag based on the mean order number

In [7]:
# Create a new column that shows the max-order per customer
df_ords_prods_merge_grp['max_order'] = df_ords_prods_merge_grp.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
# View the first fifteen rows of the dataframe
df_ords_prods_merge_grp.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                      0.0        False         196                  4   
9                     30.0        False         196                  1   
10                     7.0        False         196                  2   
11                     8.0        False         196                  6   
12                     7.0        False         196                  1   
13                     7.0        False         196                  1   
14                    14.0        False         196                  1   

    reordered product_name  aisle_id  department_id  prices _merge  \
0           0         Soda        77              7     9.0   both   
1           1         Soda        77              7     9.0   both   
2           1         Soda        77              7     9.0   both   
3           1         Soda        77              7     9.0   both   
4           1         Soda        77              7     9.0   both   
5           1         Soda        77              7     9.0   both   
6           1         Soda        77              7     9.0   both   
7           1         Soda        77              7     9.0   both   
8           1         Soda        77              7     9.0   both   
9           1         Soda        77              7     9.0   both   
10          0         Soda        77              7     9.0   both   
11          1         Soda        77              7     9.0   both   
12          1         Soda        77              7     9.0   both   
13          1         Soda        77              7     9.0   both   
14          1         Soda        77              7     9.0   both   

      price_range_loc     busiest_day     Busiest_days  \
0   Mid-range product  Regularly busy     Regular days   
1   Mid-range product  Regularly busy  Least busy days   
2   Mid-range product  Regularly busy  Least busy days   
3   Mid-range product      Least busy  Least busy days   
4   Mid-range product      Least busy  Least busy days   
5   Mid-range product  Regularly busy     Regular days   
6   Mid-range product  Regularly busy     B

In [9]:
# Create a loyalty flag for the three different customer groups using the iloc function
# Loyal customers with maximum orders over 40
df_ords_prods_merge_grp.loc[df_ords_prods_merge_grp['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
# Regular customers with maximum orders over 10 and less than or equal to 40
df_ords_prods_merge_grp.loc[(df_ords_prods_merge_grp['max_order'] <= 40) & (df_ords_prods_merge_grp['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
# New customers with maximum orders equla to or less than 10
df_ords_prods_merge_grp.loc[df_ords_prods_merge_grp['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
# View the first thirty rows of the dataframe
df_ords_prods_merge_grp.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                      0.0        False         196                  4   
9                     30.0        False         196                  1   
10                     7.0        False         196                  2   
11                     8.0        False         196                  6   
12                     7.0        False         196                  1   
13                     7.0        False         196                  1   
14                    14.0        False         196                  1   
15                     6.0        False         196                  3   
16                     8.0        False         196                 15   
17                     6.0        False         196                  1   
18                     6.0        False         196                  2   
19                     7.0        False         196                 17   
20                     8.0     

In [11]:
# View the first 60 rows of the 'user_id', 'loyalty_flag', 'order_number' columns
df_ords_prods_merge_grp[['user_id', 'loyalty_flag', 'max_order']].head(60)

user_id      loyalty_flag  max_order
0         1      New customer         10
1         1      New customer         10
2         1      New customer         10
3         1      New customer         10
4         1      New customer         10
5         1      New customer         10
6         1      New customer         10
7         1      New customer         10
8         1      New customer         10
9         1      New customer         10
10       15  Regular customer         22
11       15  Regular customer         22
12       15  Regular customer         22
13       15  Regular customer         22
14       15  Regular customer         22
15       19      New customer          9
16       19      New customer          9
17       19      New customer          9
18       21  Regular customer         33
19       31  Regular customer         20
20       31  Regular customer         20
21       43  Regular customer         11
22       43  Regular customer         11
23       52  Regular customer         27
24       52  Regular customer         27
25       52  Regular customer         27
26       52  Regular customer         27
27       52  Regular customer         27
28       52  Regular customer         27
29       52  Regular customer         27
30       52  Regular customer         27
31       52  Regular customer         27
32       52  Regular customer         27
33       52  Regular customer         27
34       52  Regular customer         27
35       52  Regular customer         27
36       52  Regular customer         27
37       67  Regular customer         24
38       67  Regular customer         24
39       67  Regular customer         24
40       67  Regular customer         24
41       67  Regular customer         24
42       67  Regular customer         24
43       67  Regular customer         24
44       67  Regular customer         24
45       67  Regular customer         24
46       67  Regular customer         24
47       67  Regular customer         24
48       67  Regular customer         24
49       67  Regular customer         24
50       67  Regular customer         24
51       67  Regular customer         24
52       67  Regular customer         24
53       67  Regular customer         24
54       67  Regular customer         24
55       67  Regular customer         24
56       81      New customer          7
57       81      New customer          7
58       82  Regular customer         19
59       82  Regular customer         19

In [12]:
# View the value counts for the loyalty_flag
df_ords_prods_merge_grp['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [13]:
# View the shape of the dataframe df_ords_prods_merge_grp
df_ords_prods_merge_grp.shape

(32404859, 21)

5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

###  Calculate the statistics for the prices column grouped by the loyalty column 

In [14]:
# Calculate the statistics for the prices column grouped by the loyalty_flag column rounded to two decimal places
df_ords_prods_merge_grp.groupby('loyalty_flag').agg({'prices': ['mean', 'median', 'min', 'max']}).round(2)

prices                     
                   mean median  min      max
loyalty_flag                                
Loyal customer    10.39    7.4  1.0  99999.0
New customer      13.29    7.4  1.0  99999.0
Regular customer  12.50    7.4  1.0  99999.0

#### As a group New customers spend the most (mean 13.29) and the Loyal customer group the least (mean 10.39)., while the mean for the Regular customer group is 12.50.

6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

### Create a spending flag for the two different customer_average_prices groups using the iloc function

In [15]:
# Create a new column that shows the average price per customer (user_id) rounded to two decimal places
df_ords_prods_merge_grp['customer_average_prices'] = df_ords_prods_merge_grp.groupby(['user_id'])['prices'].transform(np.mean).round(2)

In [16]:
# View the first fifteen rows of the dataframe
df_ords_prods_merge_grp.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                      0.0        False         196                  4   
9                     30.0        False         196                  1   
10                     7.0        False         196                  2   
11                     8.0        False         196                  6   
12                     7.0        False         196                  1   
13                     7.0        False         196                  1   
14                    14.0        False         196                  1   

    reordered  ... department_id  prices  _merge    price_range_loc  \
0           0  ...             7     9.0    both  Mid-range product   
1           1  ...             7     9.0    both  Mid-range product   
2           1  ...             7     9.0    both  Mid-range product   
3           1  ...             7     9.0    both  Mid-range product   
4           1  ...             7     9.0    both  Mid-range product   
5           1  ...             7     9.0    both  Mid-range product   
6           1  ...             7     9.0    both  Mid-range product   
7           1  ...             7     9.0    both  Mid-range product   
8           1  ...             7     9.0    both  Mid-range product   
9           1  ...             7     9.0    both  Mid-range product   
10          0  ...             7     9.0    both  Mid-range product   
11          1  ...             7     9.0    both  Mid-range product   
12          1  ...             7     9.0    both  Mid-range product   
13          1  ...             7     9.0    both  Mid-range product   
14          1  ...             7     9.0    both  Mid-range product   

       busiest_day     Busiest_days busiest_period_of_the_day max_order  \
0   Regularly busy     Regular days            Average orders        10   
1   Regularly busy  Least busy days            Average orders        10   
2   Regularly busy  Least busy days               Most orders        10   
3       Least busy  Least busy days            Average orders        10   
4       Least busy  Least busy days               Most ord

In [17]:
# Create a spending flag for the two different customer_average_prices groups using the iloc function
# Low spenders with an average prices of less than 10
df_ords_prods_merge_grp.loc[df_ords_prods_merge_grp['customer_average_prices'] < 10, 'spending_flag'] = 'Low spender'
# High spenders with an average prices of 10 or above
df_ords_prods_merge_grp.loc[df_ords_prods_merge_grp['customer_average_prices'] >= 10, 'spending_flag'] = 'High spender'

In [18]:
# View the first thirty rows of the dataframe
df_ords_prods_merge_grp.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                      0.0        False         196                  4   
9                     30.0        False         196                  1   
10                     7.0        False         196                  2   
11                     8.0        False         196                  6   
12                     7.0        False         196                  1   
13                     7.0        False         196                  1   
14                    14.0        False         196                  1   
15                     6.0        False         196                  3   
16                     8.0        False         196                 15   
17                     6.0        False         196                  1   
18                     6.0        False         196                  2   
19                     7.0        False         196                 17   
20                     8.0     

In [19]:
# View the first 60 rows of the 'user_id', 'spending_flag', 'customer_average_prices' columns
df_ords_prods_merge_grp[['user_id', 'spending_flag', 'customer_average_prices']].head(60)

user_id spending_flag  customer_average_prices
0         1   Low spender                     6.37
1         1   Low spender                     6.37
2         1   Low spender                     6.37
3         1   Low spender                     6.37
4         1   Low spender                     6.37
5         1   Low spender                     6.37
6         1   Low spender                     6.37
7         1   Low spender                     6.37
8         1   Low spender                     6.37
9         1   Low spender                     6.37
10       15   Low spender                     3.98
11       15   Low spender                     3.98
12       15   Low spender                     3.98
13       15   Low spender                     3.98
14       15   Low spender                     3.98
15       19   Low spender                     8.37
16       19   Low spender                     8.37
17       19   Low spender                     8.37
18       21   Low spender                     7.01
19       31   Low spender                     8.14
20       31   Low spender                     8.14
21       43   Low spender                     7.44
22       43   Low spender                     7.44
23       52   Low spender                     6.36
24       52   Low spender                     6.36
25       52   Low spender                     6.36
26       52   Low spender                     6.36
27       52   Low spender                     6.36
28       52   Low spender                     6.36
29       52   Low spender                     6.36
30       52   Low spender                     6.36
31       52   Low spender                     6.36
32       52   Low spender                     6.36
33       52   Low spender                     6.36
34       52   Low spender                     6.36
35       52   Low spender                     6.36
36       52   Low spender                     6.36
37       67   Low spender                     7.66
38       67   Low spender                     7.66
39       67   Low spender                     7.66
40       67   Low spender                     7.66
41       67   Low spender                     7.66
42       67   Low spender                     7.66
43       67   Low spender                     7.66
44       67   Low spender                     7.66
45       67   Low spender                     7.66
46       67   Low spender                     7.66
47       67   Low spender                     7.66
48       67   Low spender                     7.66
49       67   Low spender                     7.66
50       67   Low spender                     7.66
51       67   Low spender                     7.66
52       67   Low spender                     7.66
53       67   Low spender                     7.66
54       67   Low spender                     7.66
55       67   Low spender                     7.66
56       81   Low spender                     8.17
57       81   Low spender                     8.17
58       82   Low spender                     6.44
59       82   Low spender                     6.44

In [20]:
# View the value counts for the spending_flag
df_ords_prods_merge_grp['spending_flag'].value_counts(dropna = False)

Low spender     31769965
High spender      634894
Name: spending_flag, dtype: int64

In [21]:
# View the shape of the dataframe df_ords_prods_merge_grp
df_ords_prods_merge_grp.shape

(32404859, 23)

7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

### Create a frequency flag for the two different customer_median_prior_order_days groups using the iloc function

In [22]:
# Create a new column that shows the median days_since_prior_order per user_id (customer)
df_ords_prods_merge_grp['customer_median_prior_order_days'] = df_ords_prods_merge_grp.groupby(['user_id'])['days_since_prior_order'].transform(np.mean).round(0)

In [23]:
# View the first fifteen rows of the dataframe
df_ords_prods_merge_grp.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                      0.0        False         196                  4   
9                     30.0        False         196                  1   
10                     7.0        False         196                  2   
11                     8.0        False         196                  6   
12                     7.0        False         196                  1   
13                     7.0        False         196                  1   
14                    14.0        False         196                  1   

    reordered  ... _merge    price_range_loc     busiest_day     Busiest_days  \
0           0  ...   both  Mid-range product  Regularly busy     Regular days   
1           1  ...   both  Mid-range product  Regularly busy  Least busy days   
2           1  ...   both  Mid-range product  Regularly busy  Least busy days   
3           1  ...   both  Mid-range product      Least busy  Least busy days   
4           1  ...   both  Mid-range product      Least busy  Least busy days   
5           1  ...   both  Mid-range product  Regularly busy     Regular days   
6           1  ...   both  Mid-range product  Regularly busy     Busiest days   
7           1  ...   both  Mid-range product  Regularly busy     Busiest days   
8           1  ...   both  Mid-range product  Regularly busy     Busiest days   
9           1  ...   both  Mid-range product      Least busy  Least busy days   
10          0  ...   both  Mid-range product  Regularly busy     Busiest days   
11          1  ...   both  Mid-range product  Regularly busy     Regular days   
12          1  ...   both  Mid-range product  Regularly busy     Regular days   
13          1  ...   both  Mid-range product  Regularly busy     Busiest days   
14          1  ...   both  Mid-range product  Regularly busy     Busiest days   

   busiest_period_of_the_day max_order      loyalty_flag  \
0             Average orders        10      New customer   
1             Average orders        10      New customer   
2                Most orders        10      New customer   
3             Average orders     

In [24]:
# Create a frequency flag for the three different customer_median_prior_order_days groups using the iloc function
# Non frequent customers with the median “days_since_prior_order” greater than 20
df_ords_prods_merge_grp.loc[df_ords_prods_merge_grp['customer_median_prior_order_days'] > 20, 'frequency_flag'] = 'Non frequent customer'
# Regular customers with the median “days_since_prior_order” higher than 10 and lower than or equal to 20
df_ords_prods_merge_grp.loc[(df_ords_prods_merge_grp['customer_median_prior_order_days'] <= 20)  & (df_ords_prods_merge_grp['customer_median_prior_order_days'] > 10), 'frequency_flag'] = 'Regular Customer'
# Frequent customers with the median “days_since_prior_order” equal to or lower than 10
df_ords_prods_merge_grp.loc[df_ords_prods_merge_grp['customer_median_prior_order_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [25]:
# View the first thirty rows of the dataframe
df_ords_prods_merge_grp.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                      0.0        False         196                  4   
9                     30.0        False         196                  1   
10                     7.0        False         196                  2   
11                     8.0        False         196                  6   
12                     7.0        False         196                  1   
13                     7.0        False         196                  1   
14                    14.0        False         196                  1   
15                     6.0        False         196                  3   
16                     8.0        False         196                 15   
17                     6.0        False         196                  1   
18                     6.0        False         196                  2   
19                     7.0        False         196                 17   
20                     8.0     

In [26]:
# To view the first 60 rows of the 'user_id', 'frequencyy_flag', 'customer_median_prior_order_days' columns
df_ords_prods_merge_grp[['user_id', 'frequency_flag', 'customer_median_prior_order_days']].head(60)

user_id         frequency_flag  customer_median_prior_order_days
0         1       Regular Customer                              20.0
1         1       Regular Customer                              20.0
2         1       Regular Customer                              20.0
3         1       Regular Customer                              20.0
4         1       Regular Customer                              20.0
5         1       Regular Customer                              20.0
6         1       Regular Customer                              20.0
7         1       Regular Customer                              20.0
8         1       Regular Customer                              20.0
9         1       Regular Customer                              20.0
10       15       Regular Customer                              11.0
11       15       Regular Customer                              11.0
12       15       Regular Customer                              11.0
13       15       Regular Customer                              11.0
14       15       Regular Customer                              11.0
15       19      Frequent customer                               8.0
16       19      Frequent customer                               8.0
17       19      Frequent customer                               8.0
18       21      Frequent customer                              10.0
19       31      Frequent customer                               6.0
20       31      Frequent customer                               6.0
21       43       Regular Customer                              11.0
22       43       Regular Customer                              11.0
23       52      Frequent customer                               9.0
24       52      Frequent customer                               9.0
25       52      Frequent customer                               9.0
26       52      Frequent customer                               9.0
27       52      Frequent customer                               9.0
28       52      Frequent customer                               9.0
29       52      Frequent customer                               9.0
30       52      Frequent customer                               9.0
31       52      Frequent customer                               9.0
32       52      Frequent customer                               9.0
33       52      Frequent customer                               9.0
34       52      Frequent customer                               9.0
35       52      Frequent customer                               9.0
36       52      Frequent customer                               9.0
37       67      Frequent customer                               8.0
38       67      Frequent customer                               8.0
39       67      Frequent customer                               8.0
40       67      Frequent customer                               8.0
41       67      Frequent customer                               8.0
42       67      Frequent customer                               8.0
43       67      Frequent customer                               8.0
44       67      Frequent customer                               8.0
45       67      Frequent customer                               8.0
46       67      Frequent customer                               8.0
47       67      Frequent customer                               8.0
48       67      Frequent customer                               8.0
49       67      Frequent customer                               8.0
50       67      Frequent customer                               8.0
51       67      Frequent customer                               8.0
52       67      Frequent customer                               8.0
53       67      Frequent customer                               8.0
54       67      Frequent customer                               8.0
55       67      Frequent customer                               8.0
56       81  Non frequent customer                              26.0
57

In [27]:
# View the value counts for the frequency_flag
df_ords_prods_merge_grp['frequency_flag'].value_counts(dropna = False)

Frequent customer        17495801
Regular Customer         11812857
Non frequent customer     3096196
NaN                             5
Name: frequency_flag, dtype: int64

#### Exploring the NaN values and creating a frequency flag for them

In [28]:
# View the value counts for the NaN values in the frequency_flag
df_nan_grp = df_ords_prods_merge_grp[df_ords_prods_merge_grp['frequency_flag'].isnull() == True]

In [29]:
df_nan_grp.head()

order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
13645692                 17                     NaN         True       10749   
17251990                 17                     NaN         True       33401   
17622767                 17                     NaN         True       23695   
24138593                 17                     NaN         True       21334   
25880002                 17                     NaN         True       22198   

          add_to_cart_order  reordered  ...    price_range_loc  busiest_day  \
13645692                  3          0  ...  Mid-range product  Busiest day   
17251990                  6          0  ...  Mid-range product  Busiest day   
17622767                  2          0  ...  Low-range product  Busiest day   
24138593                  5          0  ...  Mid-range product  Busiest day   
25880002                  1          0  ...  Low-range product  Busiest day   

          Busiest_days  busiest_period_of_the_day max_order  loyalty_flag  \
13645692  Busiest days                Most orders         1  New customer   
17251990  Busiest days                Most orders         1  New customer   
17622767  Busiest days                Most orders         1  New customer   
24138593  Busiest days                Most orders         1  New customer   
25880002  Busiest days                Most orders         1  New customer   

         customer_average_prices spending_flag  \
13645692                    7.42   Low spender   
17251990                    7.42   Low spender   
17622767                    7.42   Low spender   
24138593                    7.42   Low spender   
25880002                    7.42   Low spender   

         customer_median_prior_order_days  frequency_flag  
13645692                              NaN             NaN  
17251990                              NaN             NaN  
17622767                              NaN             NaN  
24138593                              NaN             NaN  
25880002                              NaN             NaN  

[5 rows x 25 columns]

**There are 5 NaN values for user_id 159838 who only has a first order.**

In [30]:
# Create a frequency flag 'First time customer' for the  NaN customer_median_prior_order_days group using the iloc function
# First time customers with the median “days_since_prior_order” = Nan
df_ords_prods_merge_grp.loc[df_ords_prods_merge_grp['customer_median_prior_order_days'].isnull(), 'frequency_flag'] = 'First time customer'

In [31]:
# View the First time customer entries in the dataframe
df_ords_prods_merge_grp.groupby('frequency_flag').get_group('First time customer')

order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
13645692                 17                     NaN         True       10749   
17251990                 17                     NaN         True       33401   
17622767                 17                     NaN         True       23695   
24138593                 17                     NaN         True       21334   
25880002                 17                     NaN         True       22198   

          add_to_cart_order  reordered  ...    price_range_loc  busiest_day  \
13645692                  3          0  ...  Mid-range product  Busiest day   
17251990                  6          0  ...  Mid-range product  Busiest day   
17622767                  2          0  ...  Low-range product  Busiest day   
24138593                  5          0  ...  Mid-range product  Busiest day   
25880002                  1          0  ...  Low-range product  Busiest day   

          Busiest_days  busiest_period_of_the_day max_order  loyalty_flag  \
13645692  Busiest days                Most orders         1  New customer   
17251990  Busiest days                Most orders         1  New customer   
17622767  Busiest days                Most orders         1  New customer   
24138593  Busiest days                Most orders         1  New customer   
25880002  Busiest days                Most orders         1  New customer   

         customer_average_prices spending_flag  \
13645692                    7.42   Low spender   
17251990                    7.42   Low spender   
17622767                    7.42   Low spender   
24138593                    7.42   Low spender   
25880002                    7.42   Low spender   

         customer_median_prior_order_days       frequency_flag  
13645692                              NaN  First time customer  
17251990                              NaN  First time customer  
17622767                              NaN  First time customer  
24138593                              NaN  First time customer  
25880002                              NaN  First time customer  

[5 rows x 25 columns]

In [32]:
# View the first thirty rows of the dataframe
df_ords_prods_merge_grp.head(30)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                     15.0        False         196                  1   
2                     21.0        False         196                  1   
3                     29.0        False         196                  1   
4                     28.0        False         196                  1   
5                     19.0        False         196                  1   
6                     20.0        False         196                  1   
7                     14.0        False         196                  2   
8                      0.0        False         196                  4   
9                     30.0        False         196                  1   
10                     7.0        False         196                  2   
11                     8.0        False         196                  6   
12                     7.0        False         196                  1   
13                     7.0        False         196                  1   
14                    14.0        False         196                  1   
15                     6.0        False         196                  3   
16                     8.0        False         196                 15   
17                     6.0        False         196                  1   
18                     6.0        False         196                  2   
19                     7.0        False         196                 17   
20                     8.0     

In [33]:
# View the value counts for the frequency_flag
df_ords_prods_merge_grp['frequency_flag'].value_counts(dropna = False)

Frequent customer        17495801
Regular Customer         11812857
Non frequent customer     3096196
First time customer             5
Name: frequency_flag, dtype: int64

In [34]:
# View the shape of the dataframe df_ords_prods_merge_grp
df_ords_prods_merge_grp.shape

(32404859, 25)

## 04. Exporting dataframe as pickle file 'orders_products_merged_grouped.pkl'

In [ ]:
# Export the df_merged_orders_products dataframe to csv and pkl
df_ords_prods_merge_grp.to_csv(os.path.join(path, '02_Data', '02_2_Prepared_Data', 'orders_products_merged_grouped.csv'))
df_ords_prods_merge_grp.to_pickle(os.path.join(path, '02_Data', '02_2_Prepared_Data', 'orders_products_merged_grouped.pkl'))